File: packages/semantic-cache/src/adapters/langchain.ts
import { BaseCache } from '@langchain/core/caches';
import type { Generation } from '@langchain/core/outputs';
import { AIMessage } from '@langchain/core/messages';
import { SemanticCache } from '../SemanticCache';
import type { CacheCheckOptions } from '../types';
import { sha256 } from '../utils';

export interface BetterDBSemanticCacheOptions {
  /** A pre-configured SemanticCache instance. */
  cache: SemanticCache;
  /**
   * When true, cache lookups and stores are scoped to the specific LLM
   * configuration (model, temperature, etc.). This prevents cross-model
   * cache pollution but reduces hit rates — a prompt cached against gpt-4o
   * will not hit against gpt-4o-mini even if the responses would be identical.
   *
   * The llm_string is hashed (SHA-256, first 16 hex chars) for use as a
   * Valkey TAG field. The hash is deterministic: same LLM config = same hash.
   *
   * Default: false.
   */
  filterByModel?: boolean;
}

export class BetterDBSemanticCache extends BaseCache {
  private cache: SemanticCache;
  private filterByModel: boolean;
  private initPromise: Promise<void> | null = null;

  constructor(opts: BetterDBSemanticCacheOptions) {
    super();
    this.cache = opts.cache;
    this.filterByModel = opts.filterByModel ?? false;
  }

  private async ensureInitialized(): Promise<void> {
    if (!this.initPromise) {
      this.initPromise = this.cache.initialize().catch((err) => {
        this.initPromise = null; // allow retry on transient failure
        throw err;
      });
    }
    await this.initPromise;
  }

  private modelHash(llm_string: string): string {
    // llm_string is a serialised LangChain LLM config — not human-readable.
    // Hash it to a stable, TAG-safe identifier.
    return sha256(llm_string).slice(0, 16);
  }

  async lookup(prompt: string, llm_string: string): Promise<Generation[] | null> {
    await this.ensureInitialized();
    const opts: CacheCheckOptions = {};
    if (this.filterByModel) {
      opts.filter = `@model:{${this.modelHash(llm_string)}}`;
    }
    const result = await this.cache.check(prompt, opts);
    if (!result.hit || !result.response) return null;
    // Return a ChatGeneration-shaped object with a proper AIMessage so that
    // ChatOpenAI and other chat models can unwrap it correctly on cache hit.
    // Plain { text } without a message causes "Cannot read properties of undefined"
    // when the model tries to access response.content.
    return [{ text: result.response, message: new AIMessage(result.response) } as Generation];
  }

  async update(prompt: string, llm_string: string, return_val: Generation[]): Promise<void> {
    await this.ensureInitialized();
    const text = return_val.map((g) => g.text).join('');
    if (!text) return;
    await this.cache.store(prompt, text, {
      model: this.modelHash(llm_string),
    });
  }
}

File: packages/semantic-cache-py/betterdb_semantic_cache/adapters/langchain.py
"""LangChain cache adapter for betterdb-semantic-cache.

Implements LangChain's BaseCache interface backed by SemanticCache.
Async-only — synchronous lookup()/update() raise RuntimeError.

Usage::

    from betterdb_semantic_cache.adapters.langchain import BetterDBSemanticCache
    lc_cache = BetterDBSemanticCache(cache=semantic_cache)
    llm = ChatOpenAI(model="gpt-4o", cache=lc_cache)
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional, Sequence

from ..utils import sha256

try:
    from langchain_core.caches import BaseCache
    from langchain_core.messages import AIMessage
    from langchain_core.outputs import ChatGeneration, Generation
    _LANGCHAIN_AVAILABLE = True
except ImportError:
    _LANGCHAIN_AVAILABLE = False
    BaseCache = object  # type: ignore[assignment,misc]
    Generation = Any  # type: ignore[misc,assignment]
    ChatGeneration = Any  # type: ignore[misc,assignment]

if TYPE_CHECKING:
    from ..semantic_cache import SemanticCache


class BetterDBSemanticCache(BaseCache):
    """LangChain BaseCache backed by SemanticCache.

    This cache is async-only. The synchronous ``lookup``/``update`` methods
    raise ``RuntimeError`` — use an async LangChain invocation (``ainvoke`` /
    ``astream``) to avoid hitting them.

    Args:
        cache: A pre-configured SemanticCache instance.
        filter_by_model: When True, cache lookups are scoped to the LLM config.
            Prevents cross-model cache pollution but reduces hit rates.
            Default: False.
    """

    def __init__(self, cache: "SemanticCache", *, filter_by_model: bool = False) -> None:
        if not _LANGCHAIN_AVAILABLE:
            raise ImportError(
                "langchain-core is required for BetterDBSemanticCache. "
                "Install it with: pip install betterdb-semantic-cache[langchain]"
            )
        super().__init__()
        self._cache = cache
        self._filter_by_model = filter_by_model

    async def _ensure_initialized(self) -> None:
        # SemanticCache.initialize() is idempotent (guarded by asyncio.Lock),
        # so calling it directly is safe and avoids the coroutine-reuse bug.
        await self._cache.initialize()

    def _model_hash(self, llm_string: str) -> str:
        return sha256(llm_string)[:16]

    # ── Async interface (primary) ──────────────────────────────────────────

    async def alookup(
        self, prompt: str, llm_string: str
    ) -> Optional[list[Any]]:
        await self._ensure_initialized()
        from ..types import CacheCheckOptions
        opts = CacheCheckOptions()
        if self._filter_by_model:
            opts.filter = f"@model:{{{self._model_hash(llm_string)}}}"
        result = await self._cache.check(prompt, opts)
        if not result.hit or not result.response:
            return None
        return [ChatGeneration(text=result.response, message=AIMessage(result.response))]

    async def aupdate(
        self, prompt: str, llm_string: str, return_val: Sequence[Any]
    ) -> None:
        await self._ensure_initialized()
        text = "".join(
            g.text if hasattr(g, "text") else (g.get("text", "") if isinstance(g, dict) else "")
            for g in return_val
        )
        if not text:
            return
        from ..types import CacheStoreOptions
        await self._cache.store(
            prompt, text,
            CacheStoreOptions(model=self._model_hash(llm_string)),
        )

    # ── Sync interface (not supported) ────────────────────────────────────

    def lookup(self, prompt: str, llm_string: str) -> Optional[list[Any]]:
        raise RuntimeError(
            "BetterDBSemanticCache is async-only. "
            "Use an async LangChain invocation (ainvoke / astream)."
        )

    def update(self, prompt: str, llm_string: str, return_val: Sequence[Any]) -> None:
        raise RuntimeError(
            "BetterDBSemanticCache is async-only. "
            "Use an async LangChain invocation (ainvoke / astream)."
        )

    def clear(self, **kwargs: Any) -> None:
        raise RuntimeError(
            "BetterDBSemanticCache is async-only. Use aclear() instead."
        )

    async def aclear(self, **kwargs: Any) -> None:
        await self._cache.flush()
